projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php

```diff
@@ -10,24 +10,33 @@ class Jetpack_Widget_Conditions {
 	static $passed_template_redirect = false;
 
 	public static function init() {
-		if ( is_admin() ) {
+		global $pagenow;
+
+		if ( is_customize_preview() || 'widgets.php' === $pagenow ||
+			 // phpcs:ignore WordPress.Security.NonceVerification.Missing
+			( 'admin-ajax.php' === $pagenow && array_key_exists( 'action', $_POST ) && 'save-widget' === $_POST['action'] ) || // Saving widgets on classic widget admin.
+			0 === strpos( $_SERVER['REQUEST_URI'], '/wp-json/wp/v2/widget-types' ) // Widget editing via API in gutenberg widgets.
```

Comment:
Currently this code runs multiple times, eg
- go to `wp-admin/widgets.php` - first run
- for widgets to load `/wp-json/wp/v2/widget-types` is triggered - second run
- any consequent calls to `/wp-json/wp/v2/widget-types` - 3 more runs

Probably `sidebar_admin_setup` and `in_widget_form` need to run only once, on load. Should we add a check "is_loaded" to prevent subsequent runs?

Also, can we differentiate the old behavior (server rendered) from the new (client rendered) depending on if gutenberg widgets are supported (wp_use_widgets_block_editor())? For example, when gutenberg widgets are supported there is no point running the code for `is_customize_preview` or `'widgets.php' === $pagenow`.

Reply:
`wp_use_widgets_block_editor` was [introduced in wordpress 5.8](https://make.wordpress.org/core/2021/06/29/block-based-widgets-editor-in-wordpress-5-8/), so I guess we'd need to combine with a wp version check to cater for .org users on older versions of jetpack.

Reply:
Would the existence of `wp_use_widgets_block_editor` be enough for not using versions?

`function_exists( 'wp_use_widgets_block_editor' ) && wp_use_widgets_block_editor()`

Reply:
For gutenberg widget experience,  I have optimised this a little.
 * I have made the change to `sidebar_admin_setup` you mentioned. 
 * On the initial page load I don't register `in_widget_form` and `widget_update_callback`
   *  `widget_update_callback` won't be called as it's not an update, so it's slightly moot, but clearer to not register it.
   *  `in_widget_form` - we're not rendering on the server.

`in_widget_form` does _need_ to run more often though. The encode API calls and the calls to save the widgets actually return the edit form HTML again, and though gutenberg widget editing doesn't use the new form HTML, it feels correct from an API point of view that it includes the conditions HTML e.g. for other consumers. What do you think?

Reply:
I have not made any similar changes for the classic widget editor, I think there though it's not really a problem

Reply:
Thanks @dsas for refactoring this! I know it is tricky! 👀